datameta/api/submissions.py

```diff
@@ -23,59 +23,143 @@
+    # Collect missing files
+    errors = [ (file_id, "File not found") for file_id, db_file in db_files if db_file is None ]
+    # Collect authorization issues
+    errors += [ (file_id, "Access denied") for file_id, db_file in db_files if db_file is not None and (db_file.user_id!=auth_user.id or db_file.group_id!=auth_user.group_id) ]
+
+    # Collect metadatasets
+    db_msets = { mset_id : resource.resource_by_id(db, models.MetaDataSet, mset_id) for mset_id in request.openapi_validated.body['metaDataSetIds'] }
+    # Collect missing metadatasets
+    errors += [ (mset_id, "Metadata set not found") for mset_id, db_mset in db_msets if db_mset is None ]
+    # Collect authorization issues
+    errors += [ (mset_id, "Access denied") for mset_id, db_mset in db_msets if db_mset is not None and (db_mset.user_id!=auth_user.id or db_mset.group_id!=auth_user.group_id) ]
+
+    # If we collected any errors so far, raise 400 to avoid exposing internals
+    # about data the user should not be able to access
+    if errors:
+        messages, fields = zip(*errors)
+        raise errors.get_validation_error(messages, fields)
+
+    # Collect files with no data associated
+    errors += [ (file_id, "No data uploaded") for file_id, db_file in db_files if db_file.content_uploaded==False ]
+    # Collect files which already have other metadata associated
+    errors += [ (file_id, "Already submitted") for file_id, db_file in db_files if db_file.metadatumrecord is not None ]
+    # Collect metadatasets that were already submitted
+    errors += [ (mset_id, "Metadataset already submitted") for mset_id, db_mset in db_msets if db_mset.submission_id is not None ]
+
+    # Collect the file names of the provided files
+    fnames = { db_file.filename : db_file for db_file in db_files if db_file is not None }
+    # Make sure the filenames are unique
+    fname_counts = Counter(fnames.keys())
+    errors += [ (fname, "Filename occurs multiple times among provided files") for fname, count in fname_counts.items() if count > 1 ]
+
+    # Collect the file names referenced by the metadata sets
+    ref_fnames = { mdatrec.value : mdatrec for mset in db_msets.values() for mdatrec in mset if mdatrec.metadatum.isfile }
+    # Make sure referenced file names are unique
+    ref_fname_counts = Counter(ref_fnames.keys())
+    errors += [ (ref_fname, "Filename occurs multiple times in metadata") for ref_fname, count in ref_fname_counts.items() if count > 1 ]
+
+    # Make sure the files' filenames and the referenced filenames match
+    errors += [ (fname, "File included without reference in metadata") for fname in fnames.keys() if fname not in ref_fnames.keys() ]
+    errors += [ (ref_fname, "File referenced in metadata but not provided") for ref_fname in ref_fnames.keys() if ref_fname not in fnames ]
+
+    raise NotImplementedError()
+
+    # TODO
+    # - Run individual metadata validation. Q: We now have to associate the message not only with a field but also with a mset
```

Comment:
(Refering to 96) Yes that's true. If you don't like to include the mset ID in the "field", too, we could also add a "contextId" property to the ValidationError scheme?